src/react/useQueries.ts

```diff
@@ -1,11 +1,98 @@
 
 import { notifyManager } from '../core/notifyManager'
 import { QueriesObserver } from '../core/queriesObserver'
 import { useQueryClient } from './QueryClientProvider'
 import { UseQueryOptions, UseQueryResult } from './types'
 
-export function useQueries(queries: UseQueryOptions[]): UseQueryResult[] {
+type GetOptions<T extends any> =
+  // Map params from object {queryFnData: TQueryFnData, error: TError, data: TData}
+  T extends {
+    queryFnData: infer TQueryFnData
+    error?: infer TError
+    data: infer TData
+  }
+    ? UseQueryOptions<TQueryFnData, TError, TData>
+    : T extends { queryFnData: infer TQueryFnData; error?: infer TError }
+    ? UseQueryOptions<TQueryFnData, TError>
+    : T extends { data: infer TData; error?: infer TError }
+    ? UseQueryOptions<unknown, TError, TData>
+    : // Map params from tuple [TQueryFnData, TError, TData]
+    T extends [infer TQueryFnData, infer TError, infer TData]
+    ? UseQueryOptions<TQueryFnData, TError, TData>
+    : T extends [infer TQueryFnData, infer TError]
+    ? UseQueryOptions<TQueryFnData, TError>
+    : T extends [infer TQueryFnData]
+    ? UseQueryOptions<TQueryFnData>
+    : // Otherwise try to infer raw argument types
+    T extends {
+        queryFn?: QueryFunction<infer X>
+        select: (data: any) => infer Y
+      }
+    ? UseQueryOptions<X, unknown, Y>
+    : T extends { queryFn?: QueryFunction<infer X> }
+    ? UseQueryOptions<X>
+    : // Fallback
+      UseQueryOptions
+
+type GetResults<T> =
+  // Map explicit type-object to results
+  T extends { queryFnData: any; error?: infer TError; data: infer TData }
+    ? UseQueryResult<TData, TError>
+    : T extends { queryFnData: infer TQueryFnData; error?: infer TError }
+    ? UseQueryResult<TQueryFnData, TError>
+    : T extends { data: infer TData; error?: infer TError }
+    ? UseQueryResult<TData, TError>
+    : // Map explicit type-tuple to results
+    T extends [any, infer TError, infer TData]
+    ? UseQueryResult<TData, TError>
+    : T extends [infer TQueryFnData, infer TError]
+    ? UseQueryResult<TQueryFnData, TError>
+    : T extends [infer TQueryFnData]
+    ? UseQueryResult<TQueryFnData>
+    : // Otherwise map inferred type to results
+    T extends { queryFn?: QueryFunction<any>; select: (data: any) => infer Y }
+    ? UseQueryResult<Y>
+    : T extends { queryFn?: QueryFunction<infer X> }
+    ? UseQueryResult<X>
+    : // Fallback
+      UseQueryResult
+
+// in case of very large array literal, revert to UseQueryOptions[]/UseQueryResult[] to avoid TS depth-limit error
+type MAXIMUM_DEPTH = 20
+
+type QueriesOptions<
+  T extends any[],
+  Result extends any[] = [],
+  Depth extends ReadonlyArray<number> = []
+> = Depth['length'] extends MAXIMUM_DEPTH
+  ? UseQueryOptions[]
+  : T extends []
+  ? []
+  : T extends [infer Head]
+  ? [...Result, GetOptions<Head>]
+  : T extends [infer Head, ...infer Tail]
+  ? QueriesOptions<[...Tail], [...Result, GetOptions<Head>], [...Depth, 1]>
+  : T
```

Comment:
Just to add a bit more context, why I ended up going down this recursive-array-map route...

Originally I had hoped I could do something a bit more elegant with two array type parameters `TQueryFnData[]` and `TData[]` mapped by a single operation, but ran into a brick wall as I can't index them with a single key. (or if I try and force it, [it breaks the tuple magic](https://stackoverflow.com/questions/64138789/why-are-typescript-mapped-tuple-types-behaving-differently-when-supplying-generi))

Idea 1:
```ts
// ERROR: "E cannot be used to index TData"
type QueriesOptions<TQueryFnData, TData> = { [E in keyof TQueryFnData]: InferredQueryOptions<TQueryFnData[E], TData[E]> }

useQueries<TQueryFnData extends unknown[], TData extends unknown[]>(queries: [...QueriesOptions<TQueryFnData, TData>])
```

Idea 2
```ts
// this can be made to work, but QueriesOptions is no longer an array-type
type QueriesOptions<TQueryFnData, TData> = keyof TData extends keyof TQueryFnData ? { [E in keyof TQueryFnData]: InferredQueryOptions<TQueryFnData[E], TData[E]> } : never

// ERROR: "a rest parameter must be of an array type"
useQueries<TQueryFnData extends unknown[], TData extends unknown[]>(queries: [...QueriesOptions<TQueryFnData, TData>])
```